src/db/connection.rs

```diff
@@ -1,15 +1,41 @@
+use crate::db::models::Permission;
+use crate::db::models::User;
 use crate::db::sql_types::*;
+use crate::types::Token;
+use crate::utils::*;
 use diesel::pg::PgConnection;
 use rocket::fairing::{AdHoc, Fairing};
 use rocket::logger::error;
 use rocket_contrib::databases::r2d2::{Pool, PooledConnection};
 use rocket_contrib::databases::{database_config, Poolable};
 
+pub enum AuthState {
+    Valid {
+        user: User,
+        permissions: Vec<Permission>,
+    },
+    Invalid {
+        user: User,
+    },
+    Anon,
+}
+
+#[derive(Debug, Fail)]
+pub enum AuthError {
+    #[fail(display = "You are not authorized to {:?} {}", action, resource)]
+    NoPermission {
+        action: ActionType,
+        resource: String,
+    },
+    #[fail(display = "No associated user to autheticate with")]
```

Comment:
really nickpicky but authenticate is misspelled

Reply:
ahhhh ok one sec

Reply:
ok fixed!